Softeq.XToolkit.PushNotifications/PushTokenSynchronizer.cs

```diff
@@ -0,0 +1,161 @@
+        }
+
+        /// <inheritdoc />
+        public Task SynchronizeTokenIfNeededAsync()
+        {
+            if (!IsTokenRegisteredInSystem && IsTokenSavedOnServer)
+            {
+                return UnregisterFromRemotePushNotificationsAsync();
+            }
+
+            if (IsTokenRegisteredInSystem && !IsTokenSavedOnServer)
+            {
+                return DoSendTokenToServer(_pushTokenStorageService.PushToken);
+            }
+
+            return Task.CompletedTask;
+        }
+
+        /// <inheritdoc />
+        public async Task UnregisterFromRemotePushNotificationsAsync()
+        {
+            IsTokenRegisteredInSystem = false;
+
+            if (!IsTokenSavedOnServer)
+            {
+                return;
+            }
+
+            var tokenRemovedFromServer = await _remotePushNotificationsService
+                .RemovePushNotificationsToken(_pushTokenStorageService.PushToken)
+                .ConfigureAwait(false);
+            if (tokenRemovedFromServer)
+            {
+                IsTokenSavedOnServer = false;
+            }
+        }
+
+        /// <inheritdoc />
+        public void OnTokenChanged(string token)
+        {
+            if (_pushTokenStorageService.PushToken == token)
+            {
+                return;
+            }
+
+            if (string.IsNullOrWhiteSpace(token))
+            {
+                OnRegisterFailedInternalAsync().FireAndForget(_logger);
+            }
+            else
+            {
+                OnRegisterSuccessInternalAsync(token).FireAndForget(_logger);
+            }
+        }
+
+        /// <inheritdoc />
+        public async Task OnRegisterFailedInternalAsync()
+        {
+            await UnregisterFromRemotePushNotificationsAsync()
+                .ConfigureAwait(false);
+
+            _pushNotificationsHandler.OnPushRegistrationCompleted(
+                false,
+                IsTokenSavedOnServer);
+        }
+
+        private async Task OnRegisterSuccessInternalAsync(string token)
+        {
+            IsTokenRegisteredInSystem = true;
+
+            if (!IsTokenSavedOnServer)
+            {
+                _pushTokenStorageService.PushToken = token;
+
+                await DoSendTokenToServer(token)
+                    .ConfigureAwait(false);
+            }
+
+            _pushNotificationsHandler.OnPushRegistrationCompleted(
+                true,
+                IsTokenSavedOnServer);
+        }
+
+        private async Task DoSendTokenToServer(string token)
+        {
+            IsTokenSavedOnServer = await _remotePushNotificationsService
+                .SendPushNotificationsToken(token).ConfigureAwait(false);
+
+            var attemptsCount = 1;
+
+            while (!IsTokenSavedOnServer && attemptsCount < MaxAttemptsCount)
+            {
+                await Task.Delay(TokenSendRetryDelay)
+                    .ConfigureAwait(false);
+                IsTokenSavedOnServer = await _remotePushNotificationsService
```

Comment:
can we guarantee valid states in the case when sync was called twice in parallel?
```cs
SynchronizeTokenIfNeededAsync();
SynchronizeTokenIfNeededAsync();
```
`IsTokenSavedOnServer` is unpredictably in this case